check: Use --cache-dir argument

garrmcnu opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

### What is the purpose of this change? What does it change?

`check` command now uses a specific cache directory if set using the `--cache-dir` argument.

<!--
Describe the changes here, as detailed as needed.
-->

### Was the change discussed in an issue or in the forum before?

Closes #1880

<!--
Link issues and relevant forum posts here.
-->

### Checklist

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

garrmcnu committed:
check: Use --cache-dir argument

Closes #1880

fd0 reviewed (commented):
Hey, thank you very much for giving this a try! Unfortunately, the logic is not quite correct yet:

fd0 reviewed (changes requested):
Hey, thank you very much for giving this a try! Unfortunately, the logic is not quite correct yet. I'll try to explain, it's a bit more complicated than it looks at first sight ;)

`restic check` will, by default, not use the default cache. This is a safety feature: it should check the data in the repo, not the local copy. If users want, they can force restic to use the local cache by specifying `--with-cache`.

Unfortunately, not using a cache at all makes `restic check` really slow, so we're building a new cache in a temporary directory. The bug report mentioned that this is a problem for devices which may not have enough space in the temporary directory (typically `/tmp` on Linux).

So the proposal was: If `--cache-dir` is specified explicitly, use a subdir of that directory for the temporary cache instead of the system-wide temporary directory, but build a new cache from scratch and do not use the existing cache at all. When `--with-cache` is specified, the old behavior (use the existing cache) should be used as before.

What this PR implements is slightly different: when `restic check --cache-dir foo` is called, it uses the cache directory. I expect that users will have set this to their permanent cache location, so this would make `restic check` use the existing cache, which is not what we want. :)

I think we should use the `--cache-dir` given on the command-line as the input to the `ioutil.TempDir()` function, to use it as the temporary directory. If it's empty, great, use the systemd-wide temporary directory. If not, use a new subdir of that directory.

So, it'd be great if you have the motivation to try this again! Please let me know if you need any help!

garrmcnu committed:
check: Update --cache-dir argument handling based on code review comments

The temporary cache directory is created in the specified directory, or
if not specified in the default temporary directory.

fd0 reviewed (approved):
Great, looks good!

xor-gate commented:
Nice catch both, thanks for fixing!
